src/firewall/core/logger.py

```diff
@@ -134,25 +135,30 @@ def __init__(self, filename, mode="w"):
     def open(self):
         if self.fd:
             return
-        self.fd = open(self.filename, self.mode)
+        flags = os.O_CREAT|os.O_WRONLY
+        if self.mode.startswith('a'):
+            flags |= os.O_APPEND
+        self.fd = os.open(self.filename, flags, 0o640)
+        # Make sure that existing file has correct perms
+        os.fchmod(self.fd, 0o640)
         fcntl.fcntl(self.fd, fcntl.F_SETFD, fcntl.FD_CLOEXEC)
 
     def write(self, data, level, logger, is_debug=0):
         if not self.fd:
             self.open()
-        self.fd.write(data)
```

Comment:
Can you also avoid these changes by using os.fdopen() right after os.open()? I think it would also fix the remaining test failure triggered by the data.encode().